flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/PlannerConfig.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.api;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * The {@link PlannerConfig} holds parameters to configure the behavior of queries.
+ */
+@PublicEvolving
+public interface PlannerConfig {
```

Comment:
I was thinking we could try to modularize the `PlannerConfig` a bit, so that we don't need to pass along all the configuration everywhere.

We could take the example of calcite's `Wrapper` class here. And add an `unwrap` method to the `PlannerConfig`:
```
@PublicEvolving
public interface PlannerConfig {
  	@SuppressWarnings("unchecked")
	default <T extends PlannerConfig> Optional<T> unwrap(Class<T> type) {
		if (type.isInstance(this)) {
			return Optional.of((T) this);
		} else {
			return Optional.empty();
		}
	}
}
```

We could later use this method to modularize the `PlannerConfig` having sth like `CompositePlannerConfig` which would be a hashmap from class to its instance.

Reply:
What about return T directly instead of wrapping in a `Optional`.  We can return null or throw exception if the type not matched. It will allow us to write more fluent code when we are clear what the `PlannerConfig` type is, i.e. `tableConfig.getPlannerConfig().unwrap(DefaultPlannerConfig.class)`.

Reply:
As for Optional[T] or T, I'm fine with both. @dawidwys What's your opinions here?

Reply:
I am strongly for returning `Optional<T>`. This makes the contract clear. This is a user facing interface and you can never say for sure, that it contains required class. `Optional` makes it simple and clear how to handle default value: 

```
tableConfig.getPlannerConfig().unwrap(DefaultPlannerConfig.class).orElse(DefaultPlannerConfig.DEFAULT)
```

Of course you can still pass the unwrapped value internally, and unwrap it early in the stack. This way we can also modularize config. Imagine at some point we will have also e.g. `CodeGenerationConfig`. Than we can have:

```

def query(PlannerConfig config) {
   configureCalcite(config.unwarp(CalciteConfig.class).orElse(CalciteConfig.DEFAULT))
   configureCodeGen(config.unwarp(CodeGenConfig.class).orElse(CodeGenConfig.DEFAULT))
}

def configureCalcite(CalciteConfig config)

def configureCodeGeneration(CodeGenConfig config)
```